edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.IoTHub/EdgeAgentConnection.cs

```diff
@@ -139,15 +139,25 @@ async void OnConnectionStatusChanged(ConnectionStatus status, ConnectionStatusCh
         {
             try
             {
+                UpstreamProtocol protocol =
+                    this.ModuleConnection.GetModuleClient().Map(x => x.UpstreamProtocol).GetOrElse(UpstreamProtocol.Amqp);
                 Events.ConnectionStatusChanged(status, reason);
 
-                // TODO: Change this to `Device_Not_Found` after that reason is added to the C# IoT SDK
-                // and uncomment the following block to enable dynamic reprovisioning based on the
-                // new ConnectionStatusChangeReason.
-                // if (reason == ConnectionStatusChangeReason.Device_Not_Found)
-                // {
-                //    await this.deviceManager.ReprovisionDeviceAsync();
-                // }
+                // Notify the IoT Edge daemon that a device has been deprovisioned and it should check
+                // if the device has been provisioned to a different IoT hub instead.
+                // When the Amqp or AmqpWs protocol is used, the SDK returns a connection status change reason of
+                // Device_Disabled when a device is either disabled or deleted in IoT hub.
+                // For the Mqtt and MqttWs protocol however, the SDK returns a Bad_Credential status as it's not
+                // possible for IoT hub to distinguish between 'device does not exist', 'device is disabled' and
+                // 'device exists but wrong credentials were supplied' cases.
+                if ((reason == ConnectionStatusChangeReason.Device_Disabled &&
```

Comment:
Do we really need to do this? Why not simply check for reason Bad_Credential or Device_Disabled? When does AMQP throw Bad_Credential? Wouldn't we want to re-provision in that case?

Reply:
The reason I did that was to be specific about the cases where we do call Reprovision.

I'm not entirely sure when Amqp can return a Bad_Credential error. Will need to dig into the SDK or check with the SDK team.

Reply:
Took a cursory glance at the SDK and IoT hub and it seems like Hub will return UnauthorizedAccessException(s) for a bunch of cases such as SAS token expiry, cert auth failure (not really applicable here). Do you think it's a safer option to only look at Bad_Credential for Mqtt only in that case (since there's no way of Hub to distinguish between device disable/deletion/bad sas token for Mqtt only).

I can change the current condition for Device_Disabled to ignore the protocol check (for Amqp) though.